Changefeed failed with ddlSink not ready when injecting downstream network latency/partition, CDC scale out/in and capture pod failure

### What did you do?

0. TiDB cluster with 6 cdc nodes
1. Create MySQL sink changefeed 
```
cdc  cli  changefeed  create "--server=127.0.0.1:8301"  "--sink-uri=mysql://root:@downstream-tidb.cdc-testbed-tps-7542625-1-963:4000?worker-count=256
```
2. pause changefeed
3. run sysbench prepare 
```
sysbench --db-driver=mysql --mysql-host=upstream-tidb  --mysql-port=4000 --mysql-user=root --mysql-db=workload --tables=100 --table-size=50000 --create_secondary=off --debug=true --threads=32 --mysql-ignore-errors=2013,1213,1105,1205,8022,8027,8028,9004,9007,1062 oltp_write_only prepare
```
4. resume changefeed
5. run sysbench run
```
sysbench --db-driver=mysql --mysql-host=upstream-tidb  --mysql-port=4000 --mysql-user=root --mysql-db=workload --tables=100 --table-size=50000 --create_secondary=off --time=1800 --debug=true --threads=32 --mysql-ignore-errors=2013,1213,1105,1205,8022,8027,8028,9004,9007,1062 oltp_write_only run
```
6. which sysbench run in progress, do the following steps in parallel:
6.1 inject 200ms Network Latecy between CDC and downstream TiDB
6.2 Run below steps 
6.2.1 Scale in TiCDC from 6 to 2 nodes
6.2.2 Inject CDC owner Pod failure for 20s to swtich CDC owner
6.2.3 Scale out TiCDC from 2 to 6, and at the same time inject CDC to Downsteam network partition 5s.
6.2.4 restart call capture (inject Pod failure for 20s for each CDC node
6.2.5 Scale In TiCDC from 6 to 1, and at the same time inject CDC to Downstream network partition for 5s.

### What did you expect to see?

CDC changefeed should continue replication after all the ops and chaos finished, and data should be consistent

### What did you see instead?
![img_v3_02bl_25a59b49-b89c-4abb-9d5e-04877972ad9g](https://github.com/pingcap/tiflow/assets/7403864/9097e64a-7610-459f-8f0a-635ed022fece)



```
[2024/06/08 11:42:40.683 +00:00] [ERROR] [feed_state_manager.go:462] ["processor reports an error"]
[2024/06/08 11:42:40.683 +00:00] [WARN] [feed_state_manager.go:564] ["changefeed meets an error"] , "too many connections"
[2024/06/08 11:42:40.683 +00:00] [WARN] [ddl_sink.go:154] ["ddl sink initialize failed"] context canceled
[2024/06/08 11:42:40.683 +00:00] [INFO] [ddl_sink.go:282] ["owner ddl sink background loop exits"]
[2024/06/08 11:42:40.684 +00:00] [INFO] [changefeed.go:792] ["changefeed closed"]
[2024/06/08 11:42:40.733 +00:00] [WARN] [changefeed.go:294] ["an warning occurred in Owner"]  not ready
```

### Versions of the cluster

Release Version: v7.5.2
Git Commit Hash: 7c968ee228d8b775d8119a19bd42c310a0509571
Git Branch: HEAD
UTC Build Time: 2024-06-07 13:22:07
Go Version: go version go1.21.10 linux/amd64
Failpoint Build: false

--- a/cdc/owner/changefeed.go
+++ b/cdc/owner/changefeed.go
@@ -811,6 +811,18 @@ func (c *changefeed) releaseResources(ctx context.Context) {
 	c.initialized.Store(false)
 	c.isReleased = true
 
+OUT:
+	for {
+		select {
+		case err := <-c.warningCh:
+			log.Warn("drain owner warnings",
+				zap.String("namespace", c.id.Namespace),
+				zap.String("changefeed", c.id.ID),
+				zap.Error(err))
+		default:
+			break OUT
+		}
+	}
 	log.Info("changefeed closed",
 		zap.String("namespace", c.id.Namespace),
 		zap.String("changefeed", c.id.ID),

--- a/cdc/owner/changefeed.go
+++ b/cdc/owner/changefeed.go
@@ -811,6 +811,8 @@ func (c *changefeed) releaseResources(ctx context.Context) {
 	c.initialized.Store(false)
 	c.isReleased = true
 
+	// when closing a changefeed, we must clean the warningCh.
+	// otherwise, the old warning errors will be handled when the reused changefeed instance is ticked again
 OUT:
 	for {
 		select {

--- a/cdc/owner/feed_state_manager.go
+++ b/cdc/owner/feed_state_manager.go
@@ -468,6 +468,7 @@ func (m *feedStateManager) HandleWarning(errs ...*model.RunningError) {
 				"it will be failed",
 				zap.String("namespace", m.state.GetID().Namespace),
 				zap.String("changefeed", m.state.GetID().ID),
+				zap.Time("checkpointTsAdvanced", m.checkpointTsAdvanced),
 				zap.Uint64("checkpointTs", m.state.GetChangefeedStatus().CheckpointTs),
 				zap.Duration("checkpointTime", currTime.Sub(ckptTime)),
 			)

